scripts/nn/layers/graph_conv.dml

```diff
@@ -0,0 +1,262 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+/*
+ * A graph convolutional layer as presented in 'Semi-Supervised Classification with Graph Convolutional Networks'
+ * by Kipf and Welling
+ */
+
+forward = function(matrix[double] X, matrix[double] edge_index, matrix[double] edge_weight,
+                   matrix[double] W, matrix[double] b, boolean add_self_loops)
+    return (matrix[double] X_out)
+{
+    /* Forward pass of the Graph Convolutional Layer. It transforms the node feature matrix
+     * with linear weights W and then executes the message passing according to the edges.
+     * The message passing is normalized by spectral normalization, i.e. for edge (v, w) the
+     * normalization factor is 1 / sqrt(degree(v) * degree(w)).
+     *
+     * n: number of nodes.
+     * m: number of edges.
+     * f_in: number of input features per node.
+     * f_out: number of output features per node.
+     *
+     * Inputs:
+     * - X: node features, matrix of shape (n, f_in).
+     * - edge_index: directed edge list specifying the out-node (first column) and the
+     *               in-node (second column) of each edge, matrix of shape (m, 2).
+     * - edge_weight: weights of edges in edge_index, matrix of shape (m, 1).
+     *                This should be all 1s if there should be no edge weights.
+     * - W: linear weights, matrix of shape (f_in, f_out).
+     * - b: bias, matrix of shape (1, f_out).
+     * - add_self_loops: boolean that specifies whether self loops should be added.
+     *                   If TRUE new self loops will be added only for nodes that do
+     *                   not yet have a self loop. Added self loops will have weight 1.
+     *
+     * Outputs:
+     * - X_out: convolved and transformed node features, matrix of shape (n, f_out).
+     */
+    n = nrow(X)
+    m = nrow(edge_index)
+
+    # transform
+    X_hat = X %*% W
```

Comment:
This is the slow operation?

Reply:
Yes this is the slow thing. But adding the bias in the end takes just as much time.

Reply:
This matrix multiplication takes around 35% of the runtime while adding the bias in the end also takes around 35%.

Reply:
in practice adding the bias, should be very fast compared to the matrix multiplication. 
When looking at the code, it seems to me that you access indices, rather than adding vectors. 
This might be the issue.

Reply:
All things related to the index is for the convolution part and not the linear part. But, it is actually possible to do the convolution part without any indices at all since the formula for a graph convolutional layer is OUT = D^-1 * A * D^-1 * X * W + b (A: Adjacency matrix n x n, D: degree matrix n x n, X: input n x features, W: weights f_in x f_out). As you can see, to do the convolution without any indices (normalization and message passing), you need to do 3 extra matrix multiplications instead. These matrix multiplications are (in normal use cases -> n >> features) even bigger than the linear part (X*W + b) because D and A are bigger matrices than X and W. 
So, since X*W + b takes 220 seconds (only the matrix multiplaction takes around 110 seconds), not using indices to do the normalization and convolution would take way longer than 340 seconds, likely around 600 seconds.

This is also the reason why famous other libraries (PyTorch, TF) also mainly use an edge list to do the convolution part through index accessing or use a sparse matrix datatype (which is basically also an edge list) in the GCL implementation.

Reply:
You can try out the sequence of matrix multiplications the sparse representations in SystemDS are usually much faster than the other frameworks. I understand why you designed it the way you did because it does not intuitively make much sense to use the slower matrix multiplications. 

That said, the use case you currently have that runs for 220 seconds sounds a bit large because of the execution time. Can it be because the computer you use has few cores, or is the use-case you are running disproportionately large?

Reply:
That is valid point, but I believe, the fact that my implementation was 10x slower than PyTorch is not down to my implementation but down to SystemDS' performance out of the box. It might be that this implementation is indeed slow but with the linear matrix multiplication alone being 3x slower than the whole layer in PyTorch it doesn't matter how fast the convolution part will get. It won't be acceptable anyway. So trying out the sparse matrices would be very effective with the current configuration.
I would suggest running the stress test on a correct and fast configuration to see whether even the matrix multiplication alone will be faster than PyTorch. From there, it is easier to know what actually needs to be improved.

About the machines that I used for testing - I tested it on 2 different machines. One was 96-core, 190GB RAM, x86 CPU and the other was a 14-core, 16GB RAM, ARM CPU. Since both of them don't have a GPU, SystemDS should be very competitive.

Reply:
I would be surprised if our matrix multiplication is 3x slower than PyTorch's in CPU, unless PyTorch uses MKL by default or the SystemDS execution is affected by other overheads such as disk eviction. I did not compare with PyToroch but with TensorFlow, and TensorFlow was using BLAS internally.

Reply:
can it be that the memory configuration is off? what is your -Xms -Xmx arguments?

`-Xmx24g -Xms24g -Xmn2400m`

also, please when measuring did you call once? or multiple times. It is known in java that the first call is very slow.

Reply:
My -Xms -Xmx arguments were 16g, I think.

Also, I called the tests multiple times. They were always very consistent, only changing by -3 to +3 seconds.

Reply:
what i mean is not from the outside, but inside your script. 
or maybe that is what you do already?

Reply:
Sorry, the stress test file itself would probably help. The stress test is a big forward pass over 3 layers, repeated 100 times.
So, from the outside and the inside.

Reply:
ah okay, when will you be in office next time, then maybe we can talk about it?

Reply:
Sure! I will be in the office tomorrow and on friday.